File: python/sub.py
import paho.mqtt.client

class SubNode:
    def __init__(self):
        self._client = paho.mqtt.client.Client(userdata=self,
                                          protocol=paho.mqtt.client.MQTTv5,
                                          transport="tcp")
        self._client.connect("localhost")

        self._client.subscribe([("logging", paho.mqtt.client.SubscribeOptions(qos=1)),
                                ("hello", paho.mqtt.client.SubscribeOptions(qos=1))])
        self._client.message_callback_add("logging", self._callback)
        self._client.message_callback_add("hello", self._rpc_callback)
        self.a = 0

    def loop(self):
        self._client.loop_start()

    @staticmethod
    def _callback(client, userdata, message):
        userdata.a = userdata.a + 1
        print("{}".format(userdata.a))

    @staticmethod
    def _rpc_callback(client, userdata, message):
        print("Message from: {}".format(message.payload))
        userdata._client.publish(message.properties.ResponseTopic, "yo, {}".format(message.payload))

a = SubNode()
a.loop()
while True:
    pass;

File: cpp/src/sub.cpp
#include "mqtt/async_client.h"
#include <iostream>

class LogSub final : public mqtt::callback
{
public:
    LogSub(mqtt::async_client& client):
        _client{client}
    {
    }
private:
    mqtt::async_client& _client;

    void connected(const std::string& cause) override;
    void message_arrived(mqtt::const_message_ptr msg) override;
};

void LogSub::connected(const std::string& cause)
{
    std::cout << "Connected!" << std::endl;
    _client.subscribe("logging", 0);
}

void LogSub::message_arrived(mqtt::const_message_ptr msg)
{
    if(msg->get_topic() == "logging")
    {
        std::cout << "Hi " << msg->get_payload_str() << std::endl;
    }
}

int main(int argc, char *argv[])
{
    std::cout << "Starting!" << std::endl;

    mqtt::async_client client("localhost", "");

    LogSub callback(client);
    client.set_callback(callback);

    client.connect();

    while(true)
    {
    }

    return 0;
}
